src/pudl_archiver/depositors/fsspec.py

```diff
@@ -0,0 +1,272 @@
+"""Depositor implementation using any fsspec compatible file-system as storage backend."""
+
+import logging
+import traceback
+from typing import BinaryIO
+
+import aiohttp
+from pydantic import BaseModel, ConfigDict
+from upath import UPath
+
+from ..frictionless import MEDIA_TYPES, DataPackage, Resource, ResourceInfo
+from ..utils import RunSettings, compute_md5
+from .depositor import (
+    DepositionAction,
+    DepositionChange,
+    DepositorAPIClient,
+    DraftDeposition,
+    PublishedDeposition,
+    register_depositor,
+)
+
+logger = logging.getLogger(f"catalystcoop.{__name__}")
+
+
+def _resource_from_upath(path: UPath, parts: dict[str, str]) -> Resource:
+    """Create a resource from a single file with partitions.
+
+    Args:
+        file: Deposition file metadata returned by Zenodo api.
+        parts: Working partitions of current resource.
+    """
+    mt = MEDIA_TYPES[path.suffix[1:]]
+
+    return Resource(
+        name=path.name,
+        path=path.as_uri(),
+        remote_url=path.as_uri(),
+        title=path.name,
+        mediatype=mt,
+        parts=parts,
+        bytes=path.stat().st_size,
+        hash=compute_md5(path),
+        format=path.suffix,
+    )
+
+
+class Deposition(BaseModel):
+    """Represent an fsspec deposition."""
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    deposition_path: UPath
+    #: static list of files from deposition prior to any modifications
+    file_list: list[str]
+    version: str = "0.1"
+
+    @classmethod
+    def from_upath(cls, deposition_path: UPath):
+        """Construct deposition object from fsspec path to deposition."""
+        assert deposition_path.exists()
+        file_list = [
+            str(child.name)
+            for child in deposition_path.iterdir()
+            if child.name != deposition_path.name
+        ]
+        return cls(
+            deposition_path=deposition_path,
+            file_list=file_list,
+        )
+
+
+class FsspecAPIClient(DepositorAPIClient):
+    """Implement API for fsspec based depositors."""
+
+    base_path: str
+
+    @classmethod
+    async def initialize_client(
+        cls,
+        session: aiohttp.ClientSession,
+        sandbox: bool,
```

Comment:
Does `sandbox` have any meaning in this context? If not, does it need to be in the parameters here? Might be worth raising an assertion if it is passed to note that this isn't implemented for this storage backend.

Reply:
Good call, now it will raise a `NotImplementedError` if sandbox is set.